src/transport-http-client.ts

```diff
@@ -100,46 +127,104 @@ export class TransportHttpClient<BagType extends FnsBag> implements ITransport<B
         // PULL
         // Poll for a batch (array) of new envs by HTTP GET and send them to conn.handleIncomingEnvelope
         // This loop ends when conn.status is CLOSED.
-        setImmediate2(async () => {
-            while (!this.isClosed) {
-                log('----- pull thread...');
-                let sleepTime = QUICK_POLL;
+        const pull = () => {
+            log('----- pull thread...');
+
+            if (this.isClosed) return;
+            const urlToGet = url + `for/${this.deviceId}`;
+
+            const { request, cancel: cancelRequest } = fetchWithTimeout(TIMEOUT, urlToGet);
+
+            request.then(async (response) => {
+                const reschedule = this._pullReschedules.get(conn.description);
+
                 try {
-                    // fetch (with lots of checks for closure happening during a fetch)
-                    if (this.isClosed) return;
-                    const urlToGet = url + `for/${this.deviceId}`;
-                    const response = await withTimeout(TIMEOUT, fetch(urlToGet));
                     if (this.isClosed) return;
+
                     if (!response.ok) {
                         throw new RpcErrorNetworkProblem('pull thread HTTP response was not ok');
                     }
+
                     const envs = await response.json();
+
+                    if (this.isClosed) return;
+
                     if (!Array.isArray(envs)) throw new RpcError('expected an array');
                     // poll slower when there's nothing there
-                    sleepTime = envs.length >= 1 ? QUICK_POLL : SLOW_POLL;
+
+                    //sleepTime = envs.length >= 1 ? QUICK_POLL : SLOW_POLL;
+                    if (reschedule) {
+                        const nextPoll = envs.length >= 1 ? QUICK_POLL : SLOW_POLL;
+
+                        reschedule(nextPoll);
+                    }
 
                     // pass envelopes to the connection to handle one at a time
                     conn.status.set('OPEN');
                     log(`got ${envs.length} envelopes`);
                     for (const env of envs) {
-                        if (this.isClosed) return;
+                        if (this.isClosed) return () => {};
```

Comment:
why does this return an empty function, but line 150 just does a `return`?

Reply:
Leftover from a previous experiment. Changed back to `void`!